algorithms/baselines/curious_ppo/agent_curious_ppo.py

```diff
@@ -0,0 +1,211 @@
+        self.ph_next_observations = placeholder_from_space(main_observation_space(env))
+        self.prediction_curiosity_bonus = None  # updated in self.add_cm_objectives()
+
+        # Create graph for curiosity module (ICM)
+        self._cm = CuriosityModel(
+            env, self.ph_observations, self.ph_next_observations, self.ph_actions, params.stack_past_frames,
+            params.forward_fc, params=params)
+
+        # add ICM loss keys to objective function
+        self.objectives.update(self.add_cm_objectives())  # TODO: will this merging overwrite some keys?
+
+        self.add_curiosity_summaries()
+
+        self.cm_summaries = merge_summaries(collections=['cm'])
+        self.cm_step = tf.Variable(0, trainable=False, dtype=tf.int64, name='cm_step')
+
+        cm_opt = tf.train.AdamOptimizer(learning_rate=self.params.learning_rate, name='cm_opt')
+        self.train_cm = cm_opt.minimize(self.objectives.model_loss, global_step=self.cm_step)
+
+    def add_cm_objectives(self):
+        # model losses
+        forward_loss_batch = 0.5 * tf.square(self._cm.encoded_next_obs - self._cm.predicted_obs)
+        forward_loss_batch = tf.reduce_mean(forward_loss_batch, axis=1) * self._cm.feature_vector_size
+        forward_loss = tf.reduce_mean(forward_loss_batch)
+
+        bonus = self.params.prediction_bonus_coeff * forward_loss_batch
+        self.prediction_curiosity_bonus = tf.clip_by_value(bonus, -self.params.clip_bonus, self.params.clip_bonus)
+
+        inverse_loss = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(
+            logits=self._cm.predicted_actions, labels=self.ph_actions,
+        ))
+
+        cm_beta = self.params.cm_beta
+        model_loss = forward_loss * cm_beta + inverse_loss * (1.0 - cm_beta)
+        model_loss = self.params.cm_lr_scale * model_loss
+        return AttrDict(locals())
+
+    def add_curiosity_summaries(self):
+        obj = self.objectives
+        with tf.name_scope('losses'):
+            cm_scalar = partial(tf.summary.scalar, collections=['cm'])
+            cm_scalar('curiosity_forward_loss', obj.forward_loss)
+            cm_scalar('curiosity_inverse_loss', obj.inverse_loss)
+            cm_scalar('curiosity_module_loss', obj.model_loss)
+
+    def _train(self, buffer, env_steps):
+        step = self._train_actor(buffer, env_steps)
+        self._train_critic(buffer, env_steps)
+        self._train_curiosity_module(buffer, env_steps)
+        return step
+
+    def _prediction_curiosity_bonus(self, observations, actions, next_obs, goals=None):
+        bonuses = self.session.run(
+            self.prediction_curiosity_bonus,
+            feed_dict={
+                self.ph_actions: actions,
+                self.ph_observations: observations,
+                self.ph_next_observations: next_obs,
+                self.actor_critic.ph_goal_obs: goals,
+            }
+        )
+        return bonuses
+
+    def input_dict(self, buffer, start, end):
```

Comment:
This way you will pass next_obs to all training routines, even those that don't need it, am I right? This is expensive.
It's better just to call input_dict of the base class and add missing placeholders where needed